include/gunrock/cuda/launch_box.hxx

```diff
@@ -192,12 +195,37 @@ struct launch_params_dynamic_grid_t : detail::launch_params_base_t<sm_flags_> {
 
   /**
    * @brief Launch a kernel within the given launch box.
+   *
+   * @par Overview
+   * This function is a reimplementation of `std::apply`, that allows for
+   * launching cuda kernels with launch param members of the class and a context
+   * argument. It follows the "possible implementation" of `std::apply` in the
+   * C++ reference: https://en.cppreference.com/w/cpp/utility/apply.
+   *
+   * @tparam func_t The type of the kernel function being passed in.
+   * @tparam args_tuple_t The type of the tuple of arguments being passed in.
+   * @param __ Kernel function to call.
+   * @param args Tuple of arguments to be expanded as the arguments of the
+   * kernel function.
+   * @param context Reference to the context used to launch the kernel (used for
+   * the context's stream).
    */
-  template <typename func_t, typename... args_t>
-  void launch(func_t __, args_t&&... args) {
-    __<<<  // kernel function.
-        grid_dimensions, block_dimensions, shared_memory_bytes,
-        context.stream()>>>(std::forward<args_t>(args)...);
+  template <typename func_t, typename args_tuple_t>
+  void launch(func_t&& __, args_tuple_t&& args, standard_context_t& context) {
```

Comment:
@neoblizz the challenge I am still facing is what is the best way to add this function to both launch box classes when the types of the members they access are different (meaning they don't exist in the base class so we can't move it up there).

Reply:
Can you elaborate on this? Why do you need a copy of `launch` and respective functions for each `launch_params_t` class? What I mentioned earlier was not to move it UP in the base-class, but have a final class that inherits the param_t class that contains only one implementation of the `launch` or common APIs. Instead of this: 

https://github.com/gunrock/essentials/blob/781f14dc652670b4d8452dec7077ef6dfec36690/include/gunrock/cuda/launch_box.hxx#L103-L107

Reply:
My question I guess is, why does `launch_box_t` needs to `= std::conditional_t` and not just inherit? So, you could do:

```cpp
template<typename... lp_v>
struct launch_box_t : public std::conditional_t< 
     (std::tuple_size<detail::match_launch_params_t<lp_v...>>::value == 0), 
     detail::raise_not_found_error_t<void>,  // Couldn't find params for SM ver 
     std::tuple_element_t<0, detail::match_launch_params_t<lp_v...>>> {
// move launch functions, etc. here...
};

Reply:
So you're suggesting something like the inheritance diamond pattern? the two choices on either side of the diamond are to inherit static or dynamic grid dimensions?

Reply:
If we can unite all launch box types under the same struct, and conditionally inherit static or dynamic grid dimensions, then the common struct would still need template parameters for anything that _could_ be static, which just means we'd have to use a dummy grid dim template param when going for dynamic grid dimensions.

Reply:
Is it really diamond inheritance if only one of the two gets inherited? I see it as;

```
launch_box_t
    | 
    |
launch_params_t (specialized, but only pick one)
    |
    |
launch_params_base_t
```

Reply:
> If we can unite all launch box types under the same struct, and conditionally inherit static or dynamic grid dimensions, then the common struct would still need template parameters for anything that _could_ be static, which just means we'd have to use a dummy grid dim template param when going for dynamic grid dimensions.

I am not sure if this will be a nicer solution than just separating the two.

Reply:
Okay I think I finally understand what you're saying. Make the launch box inherit, and add the launch function in the launch box. Definitely not the diamond cause it doesn't inherit both.

Reply:
Yes! 🚀